### PR TITLE
Transformer test refactor

### PR DESCRIPTION
**Pull request by cowanmeg**
Moves the transformer fusion creation into a separate class so that we can call the same fusion for unit testing (validation on smaller sizes, no dropout) and benchmarking (separate script coming in later PR).

### CODE REVIEW

File: tests/cpp/multidevice_transformer.cpp [hunk 1]
```diff
@@ -0,0 +1,660 @@
+// clang-format off
```
Comment (cowanmeg) [inline tests/cpp/multidevice_transformer.cpp:1]: This file is almost directly copy paste from `test_multidevice_transformer.cpp`
There was some slight clean up in merging Forward and Forward_SP test into one parameterized function to reduce (some) duplication.

File: tests/cpp/test_multidevice_transformer.cpp [hunk 2]
```diff
@@ -13,25 +13,33 @@
 #include <fusion.h>
 #include <ops/all_ops.h>
 #include <tests/cpp/multidevice.h>
+#include <tests/cpp/multidevice_transformer.h>
 #include <tests/cpp/validator.h>
 
 namespace nvfuser {
 
-constexpr int64_t B = 2, E = 768, H = 16, S = 128;
+namespace {
```
Comment (cowanmeg) [inline tests/cpp/test_multidevice_transformer.cpp:21]: Note: Lots of removed code that went to `multipdevice_transformer.cpp`

File: tests/cpp/multidevice_transformer.cpp [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
```
Comment (wujingyue) [inline tests/cpp/multidevice_transformer.cpp:3]:
```suggestion
 * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
```
I believe you started this code in 2024.

File: tests/cpp/multidevice_transformer.cpp [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+  TensorView* grad_b;
+};
+
+// x format: [i0, i1] dtype
+// weight format: [DID(D), i2/D, i1] dtype
+// grad format: [DID(D) i0, i2/D] float or dtype
+// outputs: grad_x [i0, i1] dtype
+// grad_w [DID i2/D, i1] dtype
+// grad_b [DID i2/2] dtype
+LinearBackwardsResult linear_backwards(
```
Comment (wujingyue) [inline tests/cpp/multidevice_transformer.cpp:31]:
```suggestion
LinearBackwardsResult linearBackwards(
```
It's fine leaving it as is. For new code, I'd try to stick with the style convention.

**Review by wujingyue — APPROVED** [review]

File: tests/cpp/multidevice_transformer.cpp (hunk 4 above)
Comment (cowanmeg) [reply tests/cpp/multidevice_transformer.cpp:31]: it's a good time to do little cleanups!

**Comment by cowanmeg** [comment]
!test --diff

**Pull request merged**